Mathlib/Data/List/Destutter.lean

```diff
@@ -200,43 +199,49 @@ theorem destutter_map_iff (f : α → β) (h : ∀ a b, R a b ↔ R₂ (f a) (f
       by_cases hr : (R a a2) <;>
         simp [hr, ← destutter_cons', ih]
 
-/-- Destutter' on a relation like ≠, whose negation is an equivalence, gives the same length if
+/-- For a injective function `f`, `destutter' (·≠·)` commutes with `map f`. -/
+theorem destutter_map_inj {f : α → β} (h : Function.Injective f) [DecidableEq α] [DecidableEq β] :
+    (l.destutter (·≠·)).map f = (l.map f).destutter (·≠·) :=
+  destutter_map_iff l _ f (fun _ _ ↦ (Function.Injective.ne_iff h).symm)
+
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, gives the same length if
     the first elements are not related (¬Rα a b). --/
-theorem length_destutter'_eq [IsCoequiv α R] (hab : ¬R a b) :
+theorem length_destutter'_eq [IsEquiv α Rᶜ] (hab : ¬R a b) :
     (List.destutter' R a l).length = (List.destutter' R b l).length := by
   induction l with
   | nil => simp
   | cons c cs ih =>
-    by_cases hac : (R a c)
+    by_cases hac : (Rᶜ a c)
     case pos =>
-      have hbc : R b c := cotrans_of_neg_pos hab hac
-      simp [destutter', if_pos hac, if_pos hbc]
-    case neg =>
-      have hbc : ¬R b c := coequiv_neg_of_neg_neg hab hac
+      have hbc : ¬R b c := trans (symm hab) hac
       simpa [destutter', if_neg hac, if_neg hbc]
+    case neg =>
+      have hbc : ¬Rᶜ b c := (mt (_root_.trans hab)) hac
+      simp [destutter', if_pos (Classical.not_not.1 hac), if_pos (Classical.not_not.1 hbc)]
 
-/-- Destutter' on a relation like ≠, whose negation is an equivalence, has length
+/-- `destutter'` on a relation like ≠, whose negation is an equivalence, has length
     monotonic under List.cons --/
-theorem length_destutter'_ge_length_destutter'_cons [IsCoequiv α R] :
+theorem length_destutter'_ge_length_destutter'_cons [IsEquiv α Rᶜ] :
     (List.destutter' R a (b::l)).length ≥ (List.destutter' R b l).length := by
   cases l with
   | nil => by_cases hab : (R a b) <;> simp_all [Nat.le_succ]
   | cons c cs =>
     by_cases hab : (R a b)
     case pos => simp [destutter', if_pos hab, Nat.le_succ]
-    by_cases hac : (R a c)
+    by_cases hac : (Rᶜ a c)
```

Comment:
```suggestion
    by_cases hac : R a c
```
and swap the two branches

Reply:
This actually doesn't easily work. Using the `Rᶜ a c` is a deliberate choice because it helps the typeclass inference realize that we're talking about `Rᶜ` instead of just `R`, and lets it find the relevant `trans Rᶜ`. With `by_cases hac : R a c`, the `neg` branch gets `¬R a c` which doesn't work with the typeclass inference like `Rᶜ a c` does. Of course you can get around this but it adds several lines of noise. (Maybe there's a super-easy transparent way that I'm missing.)

Reply:
You can always do
```
obtain hac | hac : R a c ∨ Rᶜ a c := em _
```

Reply:
Alrighty, changed this